### PR TITLE
Fix ABORT_ON_WASM_EXCEPTIONS wrapper function caching to work with Safari/Firefox

### PR DESCRIPTION
**Pull request by tklajnscek**
The previous version used function references as cache keys, which works fine in Chrome, but fails in Safari/Firefox when used with WASM functions (from wasmTable) as they all toString() the same.

The cache now uses either the export name or the wasmTable index as the key.

### CODE REVIEW

**Comment by sbc100** [comment]
Wait, are you saying that if I try to make map of wasm functions they will all generate the same key in the hash table?

Is this true to for exported wasm functions as well as those retrieved using table.get()?

File: src/preamble.js [hunk 1]
```diff
@@ -678,15 +678,17 @@ function createExportWrapper(name, fixedasm) {
...
+// creating the same wrappers over and over again. For wasmTable functions the
+// keys are the indices (they can't clash with names  since identifiers can't start 
+// with digits)
 var abortWrapperCache = {};
 
 // Creates a wrapper in a closure so that each wrapper gets it's own copy of 'original'
-function makeWrapper(original) {
-  var wrapper = abortWrapperCache[original];
+function makeAbortWrapper(original, name) {
+  var wrapper = abortWrapperCache[name];
```
Comment (sbc100) [inline src/preamble.js:689]: Maybe just remove the cache?    I'm not sure how it can ever help here.. since export names are unique, as are table indexes, so we can never get a cache hit this way can we?

Comment (sbc100) [reply src/preamble.js:689]: Maybe before we could get a cache hit if we export the same function under many different names, but this case should be very rare.

Comment (tklajnscek) [reply src/preamble.js:689]: I didn't have it before, but I added it after I added wasmTable instrumentation as that gets hit a lot with the same index and I didn't want it to return a new wrapper instance every time. That would wreck the GC :)

I could move it out of `makeWrapper` into the `wasmTable.get` override only and leave the original uncached since that will now never hit the cache anyways (it would before since a lot of functions shared the same invokers).

Comment (tklajnscek) [reply src/preamble.js:689]: Just thought it was more readable if it was in `makeAbortWrapper` as it always just works even if we add a 3rd instrumentation eventually

**Comment by tklajnscek** [comment]
> Wait, are you saying that if I try to make map of wasm functions they will all generate the same key in the hash table?
> 
> Is this true to for exported wasm functions as well as those retrieved using table.get()?

Yeah if you do this:
```
var f1 = wasmTable.get(1) 
var f2 = wasmTable.get(2) 
var someMap = {}
someMap[f1] = "hey"
someMap[f2] will map to the same item and return "hey"
```

It's because the toString() of the functions is always the same for _native functions_.

So I guess it'll fail for anything from WASM. I'll double check and let you know.

File: src/preamble.js [hunk 2]
```diff
@@ -745,7 +746,7 @@ function instrumentWasmTableWithAbort() {
   // Override the wasmTable get function to return the wrappers
   var realGet = wasmTable.get;
   wasmTable.get = function(i) {
-    return makeWrapper(realGet.call(wasmTable, i));
+    return makeAbortWrapper(realGet.call(wasmTable, i), i);
```
Comment (sbc100) [inline src/preamble.js:749]: I see we might call get(3) many times and expect to get the same cached result each time.

In that case maybe just move the cache down here and only do it for table slots?      

Or you could make the second argument to `makeAbortWrapper` and only trigger caching when it is defined (and don't pass it above).

Comment (tklajnscek) [reply src/preamble.js]: Yeah, I moved it in there, it's cleaner that way :)

**Comment by tklajnscek** [comment]
So regarding functions-as-keys...

When you take a regular JS function the .toString() method returns the actual function body in Chrome/FireFox/Safari so the cache works fine.

For native functions it's a bit different
Chrome will return `"function 203() { [native code] }"` - includes the wasm function index so they're unique
FireFox and Safari return `function() { [native code] }"` - notice the lack of function index - they all compare the same

Bottom line - don't use function references as keys ever until WeakMaps are supported everywhere :)

@sbc - I think this is good to go, but not sure why those two tests never ran...

File: src/preamble.js [hunk 3]
```diff
@@ -744,8 +735,13 @@ function instrumentWasmExportsWithAbort(exports) {
 function instrumentWasmTableWithAbort() {
   // Override the wasmTable get function to return the wrappers
   var realGet = wasmTable.get;
+  var wrapperCache = {};
   wasmTable.get = function(i) {
-    return makeWrapper(realGet.call(wasmTable, i));
+    var wrapper = wrapperCache[i];
+    if(!wrapper) {
```
Comment (sbc100) [inline src/preamble.js:741]: space after `if`

**Review by sbc100 — APPROVED** [review]

File: src/preamble.js (hunk 3 above)
Comment (kripken) [inline src/preamble.js:741]:
```suggestion
    if (!wrapper) {
```

**Review by kripken — COMMENTED** [review]
Another option would be to use `new Map()`, as the `.set()/.get()` methods can work on functions, and do not stringify like objects do.

The main difference between that and this is that this PR assumes the index in the table identifies a function. So you get a different wrapper for the same function present at two indexes. Another difference is that if the table is modified, this will not work as expected, which can happen in theory if the program uses `addFunction` etc.

Overall it seems safer to write with a `Map`, unless I'm missing something.

File: src/preamble.js (hunk 3 above)
Comment (tklajnscek) [reply src/preamble.js:741]: every time... 😅

**Comment by tklajnscek** [comment]
> Another option would be to use `new Map()`, as the `.set()/.get()` methods can work on functions, and do not stringify like objects do.

Yeah that would definitely be better if we can use ES6 features - what's the official stance on that? I'm avoiding all of these while writing code for this repo. I'll gladly change it though! Let me know and I'll go for it.

> The main difference between that and this is that this PR assumes the index in the table identifies a function. So you get a different wrapper for the same function present at two indexes

I think this isn't a huge issue since it's just a minor inefficiency. 

> Another difference is that if the table is modified, this will not work as expected, which can happen in theory if the program uses `addFunction` etc.

Good point - I hadn't considered that! It's fixable with the current approach - committed a change to address this.

**Comment by kripken** [comment]
Is `Map` ES6? I think it was in ES5, even in IE11 https://caniuse.com/?search=javascript%20Map

**Comment by tklajnscek** [comment]
> Is `Map` ES6? I think it was in ES5, even in IE11 https://caniuse.com/?search=javascript%20Map

Yeah, Map, WeakMap, Set, WeakSet are all ES6. All browsers support pretty much all of ES6 since ~2017 AFAIK, except IE that's on it's deathbed anyways...

In any case, I think what I have now is fine and is ES5-only.

**Comment by tklajnscek** [comment]
Rebased to have it run the tests again since two suites got stuck.
I think this is good to go.

**Comment by kripken** [comment]
Ah, I see you handle checking `func` even without a `Map`. lgtm then.

I do think Map was in JS before ES6, btw - as caniuse shows in that link from earlier, it *is* supported in IE11, unless i'm missing something. But this code looks just as good.
